public/app/features/explore/state/query.ts

```diff
@@ -273,23 +293,27 @@ export const importQueries = (
     }
 
     let importedQueries = queries;
-    // Check if queries can be imported from previously selected datasource
-    if (sourceDataSource.meta?.id === targetDataSource.meta?.id) {
-      // Keep same queries if same type of datasource, but delete datasource query property to prevent mismatch of new and old data source instance
-      importedQueries = queries.map(({ datasource, ...query }) => query);
-    } else if (hasQueryExportSupport(sourceDataSource) && hasQueryImportSupport(targetDataSource)) {
-      const abstractQueries = await sourceDataSource.exportToAbstractQueries(queries);
-      importedQueries = await targetDataSource.importFromAbstractQueries(abstractQueries);
-    } else if (targetDataSource.importQueries) {
-      // Datasource-specific importers
-      importedQueries = await targetDataSource.importQueries(queries, sourceDataSource);
+    // If going to mixed, keep queries with source datasource
+    if (targetDataSource.name === MIXED_DATASOURCE_NAME) {
+      importedQueries = queries.map((query) => {
+        return { ...query, datasource: sourceDataSource.getRef() };
+      });
+    }
+    // If going from mixed, see what queries you keep by their individual datasources
```

Comment:
Should it also cover when switching data sources inside mixed?

Here's an example - it works with Loki to Elastic import when non-mixed is selected, but doesn't work when changing inside the mixed set.
https://user-images.githubusercontent.com/745532/186452503-24cb13e6-15b3-4108-b7d8-77156ad2b2c3.mov

Reply:
Good find! The culprit was that the core component we use has its own on change datasource function it calls when you change the query datasource ([src](https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryEditorRows.tsx#L142)). This doesnt contain explore's fancy import queries logic, so I created a param that lets you call an external function and call that as well as the normal logic.

One thing I noticed is if you have a mixed datasource and only one query, and you copy/paste the link or refresh, it will change the datasource to that query's datasource instead of keeping the mixed datasource. In my head, it makes sense to reduce the complexity of an explore session - if you are only using one datasource, just have that selected - but I can see where someone might not want that. What do you think?

Reply:
If that reduces complexity I would keep it as it is unless it proves to be a problem. Especially that in long term we will have only mixed data sources so at some point this will go away anyway.